src/sentry/integrations/slack/tasks.py

```diff
@@ -0,0 +1,111 @@
+from __future__ import absolute_import
+
+import six
+
+from django.conf import settings
+
+from sentry import http
+from sentry.utils import json
+from sentry.tasks.base import instrumented_task
+from sentry.models import Integration, Project, Rule
+
+from sentry.utils.redis import redis_clusters
+
+MEMBER_PREFIX = "@"
+CHANNEL_PREFIX = "#"
+
+LIST_TYPES = [
+    ("channels", "channels", CHANNEL_PREFIX),
+    ("groups", "groups", CHANNEL_PREFIX),
+    ("users", "members", MEMBER_PREFIX),
+]
+
+
+class RedisRuleStatus(object):
+    def __init__(self, uuid):
+        self.uuid = uuid
+
+        cluster_id = getattr(settings, "SENTRY_RULE_TASK_REDIS_CLUSTER", "default")
+        self.client = redis_clusters.get(cluster_id)
+        self._set_value("pending")
+
+    def _get_redis_key(self):
+        return u"slack-channel-task:1:{}".format(self.uuid)
+
+    def _set_value(self, status, rule_id=None):
+        value = self._format_value(status, rule_id)
+        self.client.set(self._get_redis_key(), u"{}".format(value), ex=60 * 60)
+
+    def _format_value(self, status, rule_id):
+        value = {"status": status}
+        if rule_id:
+            value["rule_id"] = six.text_type(rule_id)
+        if status == "failed":
+            # TODO(meredith): set some generic error? or pass in an error?
+            value["error"] = "some error"
+
+        return json.dumps(value)
+
+
+@instrumented_task(name="sentry.integrations.slack.search_channel_id", queue="integrations")
+def find_channel_id_for_rule(serializer, project_id, uuid, rule_id=None):
+    redis_rule_status = RedisRuleStatus(uuid)
+
+    try:
+        project = Project.objects.get(id=project_id)
+    except Project.DoesNotExist:
+        redis_rule_status._set_value("failed")
+        return
+
+    organization = project.organization
+    integration_id = None
+    channel_name = None
+
+    for action in serializer.validated_data["actions"]:
+        if action.get("workspace") and action.get("channel"):
+            integration_id = action["workspace"]
+            # we need to strip the prefix when searching on the channel name
+            channel_name = action["channel"].strip("#@")
+            break
+
+    try:
+        integration = Integration.objects.get(
+            provider="slack", organizations=organization, id=integration_id
+        )
+    except Integration.DoesNotExist:
+        redis_rule_status._set_value("failed")
+        return
+
+    token_payload = {"token": integration.metadata["access_token"]}
+    payload = dict(token_payload, **{"exclude_archived": False, "exclude_members": True})
+
+    session = http.build_session()
+    for list_type, result_name, prefix in LIST_TYPES:
+        cursor = ""
+        while cursor is not None:
+            # XXX(meredith): change limit to 1000 instead of 1
+            items = session.get(
+                "https://slack.com/api/%s.list" % list_type,
+                params=dict(payload, **{"cursor": cursor, "limit": 1}),
```

Comment:
@MeredithAnya would it be too much work to use a normal HTTP client instead of having to call `http.build_session()` then wrapping this API call in a util method we can call in both places?

Reply:
actually...hold off on this. I am wondering if the full loop can be made into a single function you can call twice with different args